sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala

```diff
@@ -291,7 +298,8 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
             val naf = if (af.children.forall(_.foldable)) {
               af
             } else {
-              patchAggregateFunctionChildren(af) { x =>
+              patchAggregateFunctionChildren(af) { x1 =>
+                val x = funcChildrenLookup.getOrElse(x1, x1)
```

Comment:
Here's one of the complications, and my solution is somewhat brittle.

When grouping by `ExpressionSet`, in the case where there are superficially different children, we don't get all of the original children in the keys of `distinctAggGroups`. This is because multiple `ExpressionSet`s may have the same baseSet but different originals, and `groupBy` chooses only one `ExpressionSet` to represent the group's key (which is what want: we want `groupBy` to group by semantically equivalent children).

However, because `distinctAggGroups` is missing some original children in its keys, `distinctAggChildAttrLookup` is also missing some original children in its keys.

To bridge this gap, I used `funcChildrenLookup`. This data structure maps each original function child to the first semantically equivalent original function child. `funcChildrenLookup` will translate the original function child into the key (hopefully) expected by `distinctAggChildAttrLookup`. The brittleness is this: this code depends, at the very least, on which `ExpressionSet` is chosen by `groupBy` as the winner.

In the [first version](https://github.com/apache/spark/compare/master...bersprockets:spark:rewritedistinct_issue_orig?expand=1) of my PR, I modified the Aggregate (if needed) so there are no superfically different function children, thus there is no complexity when performing the groupings and the patching. I find it bit more straightforward to reason about.